### ISSUE TITLE
(Trac #243) add listener-endpoint plugins

### ISSUE THREAD

**Issue by foolscap-trac** [foolscap-trac/test-foolscap-import-5#239]
| <img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?v=4" width="50" height="50"> | trac:warner reported |
|-|-|
| Trac ID | trac#243 |
| Type | enhancement |
| Created | 2015-09-22 18:48:02Z |
| Version | 0.7.0 |
| Resolution | wontfix |

As discussed in https://tahoe-lafs.org/trac/tahoe-lafs/ticket/517#comment:54 , it will help Tahoe if we can add a server-side plugin mechanism to Foolscap.

Once #204 is done, `tub.listenOn()` will be able to accept arbitrary server-endpoint descriptor strings: anything that Twisted's `endpoints.serverFromString()` can handle. This plugin will layer in front of that, so calling `tub.addServerPlugin("foo", plugin)` will make a subsequent `tub.listenOn("foo:stuff")` call get its endpoint from `plugin.get_endpoint("foo:stuff")` instead of `endpoints.serverFromString("foo:stuff")`.

If no plugin is registered for the given type, it will always fall back to `endpoints.serverFromString()`. This is different from clients, where unrecognized connection hints are ignored. The security properties for server descriptors are different (they come from the local admin, not the remote FURL provider), so we don't need to prohibit any particular types.

This gives the plugin the opportunity to include additional configuration options when constructing the endpoint. Twisted has a built-in plugin mechanism for installing endpoint-string parsers, but it is global: any extra information (like which Tor daemon to use when allocation a .onion address) would have to be provided in the descriptor string itself. For Tahoe, that's a nuisance, because we must already provide this data elsewhere for client purposes. Without plugins at the Foolscap layer, the Tahoe config would need to duplicate this data in both the client section and the `tub.port` string.

**Comment by foolscap-trac** [comment]
| <img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?v=4" width="50" height="50"> | trac:warner commented |
|-|-|

Milestone renamed
